lib/internal/modules/run_main.js

```diff
@@ -162,7 +157,34 @@ function runEntryPointWithESMLoader(callback) {
 function executeUserEntryPoint(main = process.argv[1]) {
   const resolvedMain = resolveMainPath(main);
   const useESMLoader = shouldUseESMLoader(resolvedMain);
-  if (useESMLoader) {
+
+  // We might retry as ESM if the CommonJS loader throws because the file is ESM, so
+  // try the CommonJS loader first unless we know it's supposed to be parsed as ESM.
+  let retryAsESM = false;
+  if (!useESMLoader) {
+    // Module._load is the monkey-patchable CJS module loader.
+    const { Module } = require('internal/modules/cjs/loader');
+    if (getOptionValue('--experimental-detect-module')) {
+      try {
+        Module._load(main, null, true);
```

Comment:
Why do we call this in here? Is the line 165 related?

Reply:
Line 165 is a comment that predates this PR, it just got moved into the try/catch. We’re calling this within the try/catch so that we can retry as ESM.